### DISCUSSION TITLE
Could use some further documentation about using (ironscheme contracts) and related macros

### DISCUSSION THREAD

**Discussion by bcalco** [leppie/IronScheme#112] [Discussions]
@leppie 

The wiki is good to raise awareness of features, but it's pretty Spartan in terms of detail. Some of the code provided in the `(ironscheme ...)` namespaces make sense without further elucidation. However, for serious use wrapping .NET libraries and APIs I find myself really wishing a complete guide to types and contracts and their related macros was written, as I want to use them, but it's slow going. 

Any advice about using the types and contracts namespaces correctly is welcome. I'll use this space to record specific questions as I need help with them.

**Comment by leppie** [comment]
`(ironscheme contracts)` is mostly used for R6RS compatibility. Strictly it should not be needed with good code. It is also quite primitive and not very optimized. You will see a lot of code in the base libraries trying to avoid calling a  proc decorated with it. 

That said, it is quite handy to assert arguments passed to a proc.

I just had a look at again, yeah macros are hard to grok, and I would probably take me a good 2-3 days to understand what I did here 🗡️ 

Probably best to look at examples.

Edit: You are asking about macros I wrote a good 12 years ago :D

**Comment by leppie** [comment]
This is probably a good place to use the .NET debugger. The helper class is compiled as normal proc, so it should support some form of debugging. Looking into variables might be pretty hard though.

**Comment by leppie** [comment]
Another good way of looking at this is via a .NET decompiler. It generates pretty good C# looking code.

So as an example, this is from the base libs:
```
  (define/contract (string-format fmt:string . args)
    (clr-static-call String (Format String Object[]) (->string fmt) (list->vector args)))
```

Basically if you decorate a parameter with `:`, it expects a predicate after, in this case `string?`. I think it appends the `?` and then it will lookup that proc and call it at the start of the proc.

It is a common convention in Scheme for a proc that only returns a boolean to have its name ending with `?`.

So it emits this:
```
public static object base::string-format+(object fmt, object args)
{
	object obj = #.ironscheme.predicates::string?(fmt);
	if (obj is bool && !(bool)obj)
	{
		#.ironscheme.exceptions::assertion-violation+(#.$s$string-format, "not string", Builtins.List(fmt));
	}
	return string.Format(#.ironscheme.base::->string(fmt), Helpers.RequiresArray<object>(#.ironscheme.base::list->vector(args)));
}
```

The `+` at the end of the proc just indicates that the last parameter is 'varargs', in C# terms `params object[]`. 

That is just a calling convention used by the compiler.

So you can effectively use any predicate proc to assert variables.

Example:

```
(define (is-one? x) (= x 1))

(define/contract (foo x:is-one) x)
```

Now if you call `foo` with anything but `1` as the argument, it will raise an assertion.

PS: still dont ask me to explain the macro :D

**Comment by leppie** [comment]
And for this case:
```
  (define/contract (string-append . args:string)
    (clr-static-call String 
                     (Concat String[])
                     (list->vector (map ->string args))))
```
It will check that every argument is a string.

**Reply by leppie** [reply]
That said, I am not sure why the call to `->string` is there. Looks a bit weird. Might be for mutable strings which are `StringBuilder`.

**Reply by leppie** [reply]
Like I said, good code do not need these. 

I would love to be able to have a compile mode that gets rid of these checks to run as fast as possible.
